Introducing TensorRT operator

samikama opened the pull request:
This PR introduces a new op that wraps around an highly optimized TensorRT engine and provides a seamless integration between TensorRT and TensorFlow.
- Add a TRTEngineOp that encapsulates a TensorRT executable.
- Add CreateInferenceGraph to contract a TensorRT-compilable subgraph to a TRTEngineOp.
- Update BUILD files to include new contrib package
- Add tensorflow.contrib.tensorrt python package to expose API to python

samikama committed:
Introducing TensortRT Operator to TF which can run (sub)graphs in
highly optimized TensorRT engines.  This commit is a merged version of
many commits by

   benbarsdell    <bbarsdell at nvidia.com>
   deadeyegoodwin <davidg at nvidia.com
   jjsjann123     <jiej at nvidia.com>
   samikama      <skama at  nvidia.com>

benoitsteiner committed:
Turn the op_performance_data proto lib into a header only library by default

PiperOrigin-RevId: 182621348

Signed-off-by: Jie <jiej@nvidia.com>

benoitsteiner commented on tensorflow/contrib/tensorrt/convert/inferShapes.cc:
```diff
@@ -0,0 +1,125 @@
+                               const std::vector<std::string>& output_names,
+                               ShapeMap& shapes) {
+  tensorflow::Graph g(OpRegistry::Global());
+  TF_RETURN_IF_ERROR(tensorflow::ConvertGraphDefToGraph(
```
You can also use https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/grappler/costs/graph_properties.h to infer shapes. It works directly on the GraphDef, so you don't have to create a Graph, and it's much more accurate than the ShapeRefiner

samikama replied:
Thanks @benoitsteiner ,

We had some issues with grappler but they are resolved in recent commits. We have it implemented in our working branch and will switch using grappler in the next PR.

jjsjann123 replied:
Cherry picked the grappler fix.
Updated the shape inference as well as optimization pass. Pushed to the public repo (320e07a).

benoitsteiner commented on third_party/tensorrt/LICENSE:
```diff
@@ -0,0 +1,203 @@
+Copyright 2015 The TensorFlow Authors.  All rights reserved.
```
s/2015/2018/

jjsjann123 committed:
[BUG_FIX]
  Padding bug fixes:
    Conv2D & Pad asymmetric padding index error.

yangjunpro commented:
@samikama 
Thanks for the nice work.  
Actually our team are working on the same task of integrating TensorRT with TensorFlow with the same design idea even with some similar function/variable name:). it looks that NVIDIA team run a little bit faster.

May I know have you made any benchmark with current implementation? Since with this integration TensorRT actually could leverage existing TF ops for its functionality completeness which is actually a headache for traditional TensorRT execution solution. 
But there may be some potential performance overhead between the switch of TensorRT of TF(around several hundred us to 1ms observed in our scenarios), also if there are multiple switches between TensorRT and TF,  we need to be more careful about its overhead(considering TRT->TF->TRT execution plan).

So any performance number sharing will be really helpful.

If the work is solid enough, we would like to contribute more enhancement based on PR and apply it into our internal environment and I think this will be a win-win for community, NV and Alibaba.

samikama commented:
@yangjunpro 
Thanks,
We have some preliminary measurements(@jjsjann123) for ourselves to see whether everything working or not and we see nice improvement. But we didn't go through performance in detail yet so we expect some more improvement soon. We can try to do some more measurements and get back to you.

jjsjann123 committed:
[UPDATE]
  Refactoring optimization pass and shape inference

shape inference
  Removed shape refiner and apply shape inference through
  grappler/costs/graph_properties

  Currently using static shape inference

optimization pass
  Moved Python optimization pass to c++ optimization pass

jjsjann123 committed:
[BUG_FIX]
  git patch issue fixed.
modified:   tensorflow/contrib/tensorrt/convert/convert_graph.cc

samikama committed:
Fix License date

zheng-xq commented on tensorflow/contrib/tensorrt/convert/convert_graph.cc:
```diff
@@ -0,0 +1,308 @@
+static std::unordered_set<std::string> output_nodes;
+bool IsTensorRTCandidate(const tensorflow::NodeDef& node_def) {
+  static const std::set<std::string> candidate_ops = {
+      "Identity", "Const", "Conv2D", "MaxPool", "BiasAdd", "Relu",
```
This is fine in this PR. But since the op name may change. For example: BiasAdd is after BiasAddV1. Newer versions may be added and this would easily get out of sync.

It's a good idea to split this into a registration mechanism. And then we can register the kernel to have this attribute in the kernel implementation.

A TODO item is good enough.

jjsjann123 replied:
Sure.
For this PR, current segmentation contains only the bare minimum to get basic stuff working. We need more than op type to determine whether it is TRT-compatible or not.
I'll put a TODO here and maybe we can get back to it later.

wujingyue replied:
It'd be really great to add a IfChangeThenChange lint check. This will make sure you change convert_nodes.cc when this list is changed. For example, https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/types.proto#L9

zheng-xq commented on tensorflow/contrib/tensorrt/convert/convert_graph.cc:
```diff
@@ -0,0 +1,308 @@
+  return candidate_ops.count(node_def.op());
+}
+
+void GetSubGraphIncomingEdges(tensorflow::Graph const& graph,
```
TF convention is "const Graph& graph"

zheng-xq reviewed (commented):
Some initial comments. More to come.

aaroey reviewed (commented):
Some initial comments.

wujingyue commented on configure.py:
```diff
@@ -382,13 +384,12 @@ def set_build_var(environ_cp, var_name, query_item, option_name,
-                     % (bazel_config_name, option_name))
+  # TODO(mikecase): Migrate all users of configure.py to use --config Bazel
+  # options and not to set build configs through environment variables.
+  if var=='1':
```
Several changes in this file don't follow [TF coding style](https://www.tensorflow.org/community/style_guide).

aaroey replied:
Please don't change this file at the moment, our CL for the configuration will contain the fix.

samikama replied:
@aaroey just to be clear, should we remove configure.py from the PR?

aaroey replied:
@samikama Yes, please exclude this file at the moment.

wujingyue commented on tensorflow/contrib/tensorrt/README.md:
```diff
@@ -0,0 +1,42 @@
+Using TensorRT in TensorFlow
```
This PR contains much more than the title "introducing TensorRT operator". I'd like to see a more detailed description listing specific changes. For example, 

- Add a TRTEngineOp that encapsulates a TensorRT executable. 
- Add CreateInferenceGraph to contract a TensorRT-compilable subgraph to a TRTEngineOp. 
- Add configurations that blah blah.

samikama replied:
Updated the PR message. Is this better?

wujingyue commented on tensorflow/contrib/tensorrt/convert/convert_graph.cc:
```diff
@@ -0,0 +1,308 @@
+  // AJ refactoring shape inference through grappler/GraphProperties.
+  tensorflow::grappler::GraphProperties static_graph_properties(item);
+  static_graph_properties.InferStatically(false);
+  // TF_CHECK_OK(static_graph_prop.InferStatically(false));
```
Remove all code commented out.

wujingyue commented on tensorflow/contrib/tensorrt/convert/convert_graph.cc:
```diff
@@ -0,0 +1,308 @@
+  }
+  std::unordered_map<std::string, tensorflow::Node*> node_map;
+  TF_RETURN_IF_ERROR(BuildNodeMap(graph, &node_map));
+  for (std::set<std::string> const& subgraph_node_names : segments) {
```
const std::set<std::string>&

wujingyue commented on tensorflow/contrib/tensorrt/convert/convert_graph.h:
```diff
@@ -0,0 +1,34 @@
+tensorflow::Status ConvertGraphDefToTensorRT(
+    const tensorflow::GraphDef& graph_def,
+    const std::vector<std::string>& output_names, size_t max_batch_size,
+    size_t max_workspace_size, tensorflow::GraphDef* new_graph_def);
```
Comment on `max_workspace_size`. Readers might not know what it means or even the units (in bytes or megabytes)?

wujingyue replied:
Also, who manages the work space? TF typically reserves all GPU memory ahead of time. TensorRT will run into OOM if it tries to allocate work space using cudaMalloc. @zheng-xq

samikama replied:
As of now, TensorRT doesn't have a API to use external allocator but they are aware of the situation and will try to push the api as soon as possible. Once it is available, we will use TF allocators to allocate memory. Until then, users will have to set a different allocation strategy, for example incremental allocation or fixed percentage allocation.

zheng-xq replied:
That's very unfortunate. Either we should add an external allocator, or have caller to pass in the scratch space, similar to cudnn APIs

samikama replied:
We will have a means to use TF allocators directly and TensorRT team is working on it with high priority, but right now TensorRT doesn't have any means to let users control memory other than the size of the buffer. As soon as TensorRT API becomes ready we will make another PR for integrated memory allocation.

wujingyue commented on tensorflow/contrib/tensorrt/kernels/trt_engine_op.cc:
```diff
@@ -0,0 +1,183 @@
+  infer->destroy();
+  std::stringstream oss;
+  // debug iterate through all binding instances
+  for (int i = 0; i < trt_engine_ptr_->getNbBindings(); i++) {
```
Debugging info is too level for LOG(INFO)s. Consider VLOG.

wujingyue commented on tensorflow/contrib/tensorrt/kernels/trt_engine_op.cc:
```diff
@@ -0,0 +1,183 @@
+        break;
+      case nvinfer1::DataType::kHALF:
+        LOG(INFO) << "half";
+        // buffers[bindingIndex] = (void*)input_tensor.flat<float16>().data();
```
If the current implementation doesn't support kHALF, crash instead of silently fail.

wujingyue commented on tensorflow/contrib/tensorrt/kernels/trt_engine_op.cc:
```diff
@@ -0,0 +1,183 @@
+                                                ->CudaStreamMemberHack()));
+
+  trt_context_ptr_->enqueue(nbBatch, &buffers[0], *stream, nullptr);
+  cudaStreamSynchronize(*stream);
```
Is this really needed? Synchronizing inside `OpKernel::Compute` is very expensive. @zheng-xq

aaroey replied:
Good point! We don't need to sync here.

samikama replied:
Question,
How can we ensure the output ready without synchronization before returning? Is there another mechanism?

jjsjann123 replied:
Will remove it, assuming that Tensorflow is handling execution.
We were creating our own stream in our prototype and didn't update the code here.

wujingyue replied:
@samikama tf::Tensors are ref-counted. `OpKernelContext` holds a reference of the result of an allocate_output even after `Compute` returns.

wujingyue commented on tensorflow/contrib/tensorrt/kernels/trt_engine_op.h:
```diff
@@ -0,0 +1,55 @@
+  using destroyed_ptr = std::unique_ptr<T, Destroyer<T>>;
+  destroyed_ptr<nvinfer1::ICudaEngine> trt_engine_ptr_;
+  // TODO(samikama) context should go to a resource manager!
+  destroyed_ptr<nvinfer1::IExecutionContext> trt_context_ptr_;
```
Rename `trt_context_ptr_` to `trt_execution_context_ptr_` or `trt_executable_ptr_`.

wujingyue commented on tensorflow/contrib/tensorrt/log/trt_logger.cc:
```diff
@@ -0,0 +1,56 @@
+      LOG(WARNING) << msg;
+      break;
+    }
+    case Severity::kERROR: {
```
Does `kERROR` in TensorRT land mean a fatal error? Can a program still continue with a `kERROR`? You may want to use `LOG(FATAL)` if errors are fatal.

samikama replied:
I think it will probably be better to issue the fatal outside tensorrt stack, back in the tensorflow domain. We will have more information there.

wujingyue commented on tensorflow/contrib/tensorrt/ops/trt_engine_op.cc:
```diff
@@ -0,0 +1,37 @@
+    .Attr("serialized_engine: string")
+    .Attr("input_nodes: list(string)")
+    .Attr("output_nodes: list(string)")
+    .Attr("InT: list({int8, float16, float32})")
```
If int8 and float16 support are not in this PR yet, restrict the set to only `float32` so users see the failure early during graph construction.

jjsjann123 replied:
fp16 should work out of the box. Since we haven't got the time to test and run. Will comment out for this PR. :)

samikama commented on tensorflow/contrib/tensorrt/README.md:
````diff
@@ -0,0 +1,42 @@
+the build flags during the compilation such as
+
+```
+bazel build --config=cuda --config=opt --config=tensorrt //tensorflow/tools/pip_package:build_pip_package
````
I though this in parallel to cuda, but it can be written into .tf_configure.bazelrc directly

wujingyue reviewed (changes requested):
Nice work! Find my first round of reviews inlined.

samikama commented on tensorflow/contrib/tensorrt/README.md:
```diff
@@ -0,0 +1,42 @@
+#... create and train or load model
+gdef=sess.graph.as_graph_def()
+trt_gdef=trt.CreateInferenceGraph(gdef, #original graph_def
+				  ["output"], #name of output node(s)
```
Ok. We can extract the names from the tensors and pass them to our internal function

samikama replied:
This is done in python side now, but as discussed in the meeting these need to change in upcoming PRs

samikama commented on tensorflow/BUILD:
```diff
@@ -358,6 +358,14 @@ config_setting(
     },
 )
 
+config_setting(
```
It should be used from the .tf_configure.bazelrc file. Config is enabling this. It was first introduced to conditionally construct the local_tensorrt_repository and add tensorrt to the pywrap_tensorflow. But since then we found a workaround the issue and probably can remove this.

samikama commented on tensorflow/contrib/tensorrt/convert/convert_graph.cc:
```diff
@@ -0,0 +1,308 @@
+#include "tensorflow/core/lib/core/status.h"
+#include "tensorflow/core/platform/logging.h"
+
+#define _TF_LOG_DEBUG ::tensorflow::internal::LogMessage(__FILE__, __LINE__, -1)
```
@aaroey,

They are being used, see #L234 for example. LOG(DEBUG) macro expands to _TF_LOG_DEBUG but since TF logger doesn't have a debug level, we added this macro.

samikama commented on tensorflow/contrib/tensorrt/kernels/trt_engine_op.cc:
```diff
@@ -0,0 +1,183 @@
+  OP_REQUIRES_OK(context, context->GetAttr("input_nodes", &input_nodes_));
+  OP_REQUIRES_OK(context, context->GetAttr("output_nodes", &output_nodes_));
+
+  // TODO(samikama) runtime should be taken from a resourcemanager as well.
```
Working on this right now but will probably come with another PR if you don't mind.

aaroey replied:
SG

samikama committed:
[ Fixes for PR comments]
  - removed commented code
  - moved all code to tensorflow namespace
  - removed a dangling config parameter from tensorflow/BUILD
  - fixed Copyright notice years

samikama commented on tensorflow/contrib/tensorrt/convert/convert_graph.cc:
```diff
@@ -0,0 +1,308 @@
+
+#include "tensorflow/core/grappler/costs/graph_properties.h"
+
+//------------------------------------------------------------------------------
```
Done.

deadeyegoodwin commented on tensorflow/contrib/tensorrt/segment/segment.cc:
```diff
@@ -0,0 +1,259 @@
+
+        if (CanContractEdge(out_edge, graph)) {
+          VLOG(2) << "... ... can contract";
+          contract_edges.insert(out_edge);
```
The current segmenter doesn't make any attempt to optimize globally. In the future if we have some measure of which nodes would benefit the most from tensorrt then we could implement such an algorithm. I will add a comment.

benoitsteiner replied:
We have an internal effort to try to figure out how to globally segment the graph into fused clusters. It's not ready yet, but once it is it would be interesting to interface it with TensorRT.

zheng-xq commented on tensorflow/contrib/tensorrt/README.md:
```diff
@@ -0,0 +1,42 @@
+				  max_workspace_size # max memory for TensorRT to use 
+				  )
+tf.reset_default_graph()
+tf.import_graph_def(graph_def=trt_gdef)
```
Is there a Python test for this feature yet?

jjsjann123 replied:
We currently have python scripts that does conversion and runs trt_embedded tf graph.
Should we include python test for the module?

zheng-xq commented on tensorflow/contrib/tensorrt/convert/convert_graph.cc:
```diff
@@ -0,0 +1,292 @@
+namespace convert {
+namespace {
+
+static std::unordered_set<std::string> output_nodes;
```
Can we get rid of this global variable? It could be put inside a class.

jjsjann123 replied:
temporarily passed it as arguments.

zheng-xq commented on tensorflow/contrib/tensorrt/convert/convert_graph.cc:
```diff
@@ -0,0 +1,292 @@
+                              tensorflow::EdgeSet* incoming_edges) {
+  for (int node_id : subgraph_node_ids) {
+    tensorflow::Node const* node = graph.FindNodeId(node_id);
+    LOG(DEBUG) << node->name() << " has incoming edges: ";
```
If you really want to add those logs in production code, TF convention is 

if (VLOG_IS_ON(2)) {
  VLOG(2) << ...
}

If they are for development only, please make sure to remove them for code reviews.

jjsjann123 commented on tensorflow/contrib/tensorrt/README.md:
```diff
@@ -0,0 +1,42 @@
+#... create and train or load model
+gdef=sess.graph.as_graph_def()
+trt_gdef=trt.CreateInferenceGraph(gdef, #original graph_def
+				  ["output"], #name of output node(s)
```
The grappler optimization pass we used also seems to require "fetch" node(s), which I understand refers to the output.
Please feel free to provide further insight.

jjsjann123 commented on tensorflow/contrib/tensorrt/README.md:
```diff
@@ -0,0 +1,42 @@
+#... create and train or load model
+gdef=sess.graph.as_graph_def()
+trt_gdef=trt.CreateInferenceGraph(gdef, #original graph_def
+				  ["output"], #name of output node(s)
```
This is another reason for the output nodes.
We are using the name list to exclude output nodes to included inside TRT sub_graph. Otherwise, like you said, output tensors will be renamed. We discussed about adding an identity node here, but that does not work (or we haven't figure out how) for output nodes with multiple ports.

Will put TODO here.

Commit:
improve comment describing segmentation algorithm

aaroey commented on tensorflow/contrib/tensorrt/segment/segment_test.cc:
```diff
@@ -0,0 +1,363 @@
+#include "tensorflow/core/platform/test.h"
+
+//------------------------------------------------------------------------------
+using namespace tensorflow;
```
Please [don't](https://google.github.io/styleguide/cppguide.html#Namespaces) 'using namespace xxx', make tensorflow the outermost namespace instead.

aaroey commented on tensorflow/contrib/tensorrt/shape_fn/trt_shfn.cc:
```diff
@@ -0,0 +1,123 @@
+  // debug print out engine binding;
+  std::stringstream oss;
+  for (int i = 0; i < trt_engine->getNbBindings(); i++) {
+    LOG(INFO) << "index: " << i
```
Same as above, please make them VLOG.

aaroey commented on tensorflow/contrib/tensorrt/kernels/trt_engine_op.cc:
```diff
@@ -0,0 +1,169 @@
+                                                ->CudaStreamMemberHack()));
+
+  trt_execution_context_ptr_->enqueue(nbBatch, &buffers[0], *stream, nullptr);
+  cudaStreamSynchronize(*stream);
```
Ok it seems won't work as trt manages the device memory by itself. Please add a TODO then.

jjsjann123 replied:
For the time being, TRT manages & owns the engine memory (persistent between execution). Input/output memory is owned by TF. Removing synchronization actually should work.

aaroey commented:
We've committed 76f6938bafeb81a4ca41b8dac2b9c83e1286fa95 for the build config and I'm merging the conflicts. Some difference between that commit and this PR:

1. We need to use macro GOOGLE_TENSORRT to guard all c++ files, see [tensorrt_test.cc](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/tensorrt/tensorrt_test.cc#L20) for an example.
2. Bazel macro 'if_trt' now becomes 'if_tensorrt'
3. The 'tensorrt' lib is renamed to 'nv_infer'

Meanwhile I'll try to pull in this PR locally to make sure it works, thanks.

aaroey committed:
Merge branch 'master' into tensorrt

evanluyifan commented:
I merged this PR locally with tensorflow master branch, and tried to build it with configure "build tensorflow with tensorRT support: [y/N] y"
While got this error msg: 
TensorRT library version detected from /mnt/evanlu.lyf/tensorRT/TensorRT-3.0.0/include/NvInfer.h (4           //!<.0          //!<.0          //!<) does not match TF_TENSORRT_VERSION (4.0.0). To fix this rerun configure again.
Is this PR is currently only internal for NV developers using tensorRT 4.0 version ?

cliffwoolley commented:
@evanluyifan - Thanks for the feedback.  No, this is not only for internal NV development.  The libnvinfer version numbering is simply done independently of the main TensorRT version number.  By any chance do you have the TRT 3.0.0 RC version instead of the more recent TRT 3.0.1 GA release?

wujingyue commented on tensorflow/contrib/tensorrt/convert/convert_graph.cc:
```diff
@@ -0,0 +1,292 @@
+    new tensorflow::grappler::VirtualCluster({{"/GPU:0", device_properties}});
+
+  // single machine
+  int num_cpu_cores = tensorflow::grappler::GetNumAvailableLogicalCPUCores();
```
These lines calling tensorflow::grappler::GetXYZ are useless now.

jjsjann123 replied:
Will remove it.
Having linker issue with core/grappler/clusters/singlemachine. so didn't ended up using it.

wujingyue commented on tensorflow/contrib/tensorrt/convert/convert_graph.cc:
```diff
@@ -0,0 +1,292 @@
+  // Segment the graph into subgraphs that can be converted to TensorRT
+  tensorflow::tensorrt::segment::SegmentOptions segment_options;
+  // TODO(ben,jie,sami): exclude output nodes (DISCUSS IT)
+  for (auto node : output_names) output_nodes.insert(node);
```
Style nit:

for (...) {
  ...
}

wujingyue commented on tensorflow/contrib/tensorrt/shape_fn/trt_shfn.cc:
```diff
@@ -0,0 +1,123 @@
+        if (i == 0)
+          nbBatch = c->Value(dimHandler);
+        else if (nbBatch != c->Value(dimHandler))
+          LOG(WARNING) << "!!!!!!nbBatch does not match!!!!!!";
```
Remove !s :)

wujingyue commented on tensorflow/contrib/tensorrt/shape_fn/trt_shfn.cc:
```diff
@@ -0,0 +1,123 @@
+  std::vector<string> input_nodes;
+  c->GetAttr("input_nodes", &input_nodes);
+  for (size_t i = 0; i < input_nodes.size(); i++) {
+    int index = i;
```
Nit: index is equivalent to "i". Keep one of them.

wujingyue commented on third_party/tensorrt/build_defs.bzl.tpl:
```diff
@@ -4,4 +4,4 @@ def if_tensorrt(if_true, if_false=[]):
   if %{tensorrt_is_configured}:
     return if_true
-  return if_false
+  return if_false
```
Nit: Revert this change. We do want a new line at the end.

jjsjann123 commented:
@evanluyifan the PR head is also failing on me after the master merge. As @cliffwoolley said, we are using the public TensorRT GA.
If you want to use it, try checkout commit 9384314

samikama commented:
Hi @aaroey,

76f6938 seems to break our builds as well. I believe there could be some improvements to it. Does it work for you internally?

Thanks,
Sami

aaroey commented:
@samikama I'm working on the fix.

jjsjann123 committed:
[UPDATE]
  addressing code review comments.

  TODO:
    MACRO GOOGLE_TENSORRT to guard c++ files.
    build/test -> current code works on local repo before the master merge.

jjsjann123 commented:
@aaroey Thanks!
I pushed some updates addressing the review comments. Will test it with the merge once I saw your fix :)

aaroey committed:
Fix compilation, there are still linking issues which will be fixed in a followup commit.

jjsjann123 committed:
[update]
  addressing:
  i. some naming conventions
  ii. add macro to guard cpp files

  TODO:
    cpplint

jjsjann123 committed:
[update]
  clang-format & cpplint done.

  TODO: compile and test after build fix.

aaroey committed:
Fix build (part1):
1. Changed includes of "NvInfer.h" to "tensorrt/include/NvInfer.h"
2. Remove build target "tensorrt_ops.so" (src file doesn't exist and the
   target is not used anywhere)
3. Add missing '#if GOOGLE_TENSORRT's
4. Use tf_cuda_library instead of cc_library for some targets to get the
   tf_copts naturally.
5. Revert the changes that was accidentally made (by merging with
   upstream head) from configure.py
6. Replace exception with LOG(FATAL) in
   tensorflow/contrib/tensorrt/convert/convert_nodes.cc as exception is
   not supported.
7. Revert the reinterprete_cast change in
   tensorflow/contrib/tensorrt/kernels/trt_engine_op.cc
8. Fix minor formatting and naming issues according to the style guide.

samikama committed:
Fix compilation issues with new build files

samikama committed:
Fix merge problems

aaroey commented on tensorflow/contrib/tensorrt/log/trt_logger.cc:
```diff
@@ -12,8 +12,6 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 ==============================================================================*/
-#if GOOGLE_CUDA
```
@samikama why removing this? I think it is required as if trt is not enabled we should not include the logger code.

samikama replied:
The code is not compiled with necessary macros. Moreover this rule should not trigger if trt is not enabled at all. Am I missing something?

samikama replied:
Just for my understanding, I though trt_logger is not compiled if trt is not enabled due to bazel rules. Is this not the case?

samikama replied:
Ok. Put it back. It works now since you changed the library to be a cuda library.

aaroey replied:
You're right. The logger is not compiled when trt is not enabled, but we still want it to be compilable (i.e. if I manually compile it it should not produce error) in that case. Thanks for adding back. :)

samikama commented:
Hi @aaroey,

I pushed some changes that fixes the build  for me.  Do these work for you as well?

aaroey commented:
@samikama Thanks, I also pushed some fixes. It looks like your changes conflict with mine. Would you sync to head and retry?

jjsjann123 commented on tensorflow/contrib/tensorrt/log/trt_logger.cc:
```diff
@@ -14,21 +14,23 @@ limitations under the License.
   switch (severity) {
     case Severity::kINFO: {  // mark TRT info messages as debug!
       VLOG(-1) << msg;
+    case Severity::kINFO: {  // Mark TRT info messages as debug!
```
merge conflicts not properly resolved :)

samikama replied:
@jjsjann123, yep missed that. Merge conflicts were elsewhere and didn't catch this :(

samikama commented:
Hi @aaroey,

I don't see any changes since 8e03944 . It was failing for me.

aaroey commented:
@samikama Let me fix the problem in logger. It looks like the problem is in merging: https://github.com/tensorflow/tensorflow/pull/16253/files/ae740a67bdc01b991ead6ac047c774bff4d7bc8f..fe4260f71479374aee8340b8dacdbec8102d18c3#diff-39a0a00c611f917b0c3bbc776ddd6e01R31

aaroey committed:
Resolve merge conflict

aaroey commented:
@samikama please check again.

samikama commented:
@aaroey,
Done, sorry for the missed merge issue.

samikama committed:
Remove debug log and use vlog instead

samikama commented:
@aaroey 
Just pushed another commit, removing the debug logs and replacing them with vlogs

jjsjann123 commented:
Build works for me.
But cannot import tensorrt in python
```

import tensorflow.contrib.tensorrt
2018-01-29 16:29:45.638178: F tensorflow/core/framework/op.cc:54] Non-OK-status: RegisterAlreadyLocked(op_data_factory) status: Already exists: Op with name _Arg
Aborted (core dumped)

```

samikama committed:
Update Readme and add GOOGLE_CUDA conditionals back

samikama commented on tensorflow/contrib/tensorrt/README.md:
```diff
@@ -0,0 +1,42 @@
+trt_gdef=trt.CreateInferenceGraph(gdef, #original graph_def
+				  ["output"], #name of output node(s)
+				  max_batch_size, #maximum batch size to run the inference
+				  max_workspace_size # max memory for TensorRT to use 
```
We have defaults in python code, they are not listed in README. Should we add them to there as well?

samikama committed:
Add TODO message

samikama commented:
@aaroey, @wujingyue 
Is there anything else left?

samikama commented:
It looks like, core/ops/function_ops.cc seems to be included in python bindings we create which cause static constructors to be included in our .so which in turn is causing our import to cause an abort due to multiple registration of "_Arg" system op, as observed by @jjsjann123 . At least this is my understanding. This file wasn't included before, @aaroey do you know what could have caused this change?

evanluyifan commented:
@cliffwoolley  @jjsjann123   Thansk for your response,  have you fixed the issue on building with master branch ? 
I' currently using tensorRT 3.0 RC release.  Do I need to upgrade this to 3.0.1 ?
And since this PR is based on the Public GA, why there is a TRT version check? which indicates we need tensorRT version >= 4.0 ?

BTW, I think what this PR doing is adding tensorRT support into tensorflow, is there any schedule for this?

Personally I do agree that tensorflow with tensorRT native support is a great idea. On the other hand, we also have some guys working the XLA side. I just wondering, for the inference part, especially the CNN networks (I know TRT currently could only fully support CNN networks), what would be the future of "XLA optimaztion" vs "tensorRT support"?

cliffwoolley commented:
@evanluyifan - Please do update to TensorRT 3.0 GA, available from https://developer.nvidia.com/tensorrt .  The most recent build posted there is numbered 3.0.2.

Again you don't need a "4.0" version of TensorRT - there's no such thing.  TensorRT 3 will include libnvinfer4 out of the box.

And yes this PR does provide integration of TRT into TF.  We have several future steps in mind to incrementally improve this integration and better leverage internal TF mechanisms; this is just the first step.  I cannot comment on a timeline, however.

Thanks,
Cliff

aaroey commented:
@evanluyifan I think this should work with any trt version, if it doesn't, please let me know and provide a reliable repro case. There is still some issues with the build, while we're working on the fix,  this PR should be in within a few days.

aaroey commented on tensorflow/contrib/tensorrt/kernels/trt_engine_op.cc:
```diff
@@ -0,0 +1,138 @@
+                                                ->CudaStreamMemberHack()));
+
+  // execution handled by TF since we are getting stream from TF.
+  trt_execution_context_ptr_->enqueue(num_batch, &buffers[0], *stream, nullptr);
```
@jjsjann123 'buffers' will go out of scope once this call returns, is it ok? If yes it means enqueue() will create a copy of it, manage the lifetime of the copy and delete it once the actual computation is finished, am I right? If yes, please comment.

jjsjann123 replied:
Sorry I missed your last comment about buffers run out of scope.

I checked TRT code, looks like enqueue() is not making copy of the array. I will double check with TRT team tomorrow.
Meanwhile, enqueue() provides a callback function after memory has been consumed. I can use that to maintain temporary buffer for enqueue() to avoid synchronization.

jjsjann123 replied:
Got confirmation: 
"TRT doesn't need to keep that (CPU memory holding the array of pointers) around after the kernels have been enqueued."

we should be good here.

aaroey replied:
Thanks, please add a comment then.

aaroey committed:
Fix GOOGLE_TENSORRT

aaroey committed:
Remove commented lines in configure.py

aaroey commented on tensorflow/contrib/tensorrt/BUILD:
```diff
@@ -32,6 +45,224 @@ tf_cuda_cc_test(
+
+# Library for the node-level conversion portion of TensorRT operation creation
+
+filegroup(
```
@samikama Hi Sami, what is the purpose of adding this and next rule?

samikama replied:
I remember seeing these somewhere being necessary but I couldn't find it now. I don't remember why they were needed either so I will remove them.

aaroey commented:
@samikama @jjsjann123 I'm working on fixing the build dependencies, and hopefully it would solve the problem of double linking. Just FYI.

aaroey commented on tensorflow/contrib/BUILD:
```diff
@@ -104,7 +105,9 @@ py_library(
         "//tensorflow/contrib/util:util_py",
         "//tensorflow/python:util",
-    ] + if_mpi(["//tensorflow/contrib/mpi_collectives:mpi_collectives_py"]),
+    ] + if_mpi(["//tensorflow/contrib/mpi_collectives:mpi_ops_py"])
```
@samikama why changing this line?

samikama replied:
@aaroey isn't this coming from 98ffa7b2bb ? I thought you changed it.

aaroey replied:
@samikama I thought it came from the first commit of trt?
https://github.com/samikama/tensorflow/blame/tensorrt/tensorflow/contrib/BUILD#L108

samikama replied:
Since we branched from around end of November, there was plenty of merge mistakes when we were creating the PR and had to resolve them manually. For this one, I must have taken the wrong one. We don't need any changes there.

aaroey replied:
No problem, I'll fix that along with my fixes to the build deps.

aaroey committed:
Fix the build dependencies and formatting of the code, and make sure
they follow the style conventions.

aaroey commented:
@samikama @jjsjann123 the import error is caused by grappler now:

$ bazel query 'somepath(//tensorflow/contrib/tensorrt:init_py, //tensorflow/core:functional_ops_op_lib)'
//tensorflow/contrib/tensorrt:init_py
//tensorflow/contrib/tensorrt:trt_convert_py
//tensorflow/contrib/tensorrt:wrap_conversion
//tensorflow/contrib/tensorrt:_wrap_conversion.so
//tensorflow/contrib/tensorrt:trt_conversion
//tensorflow/core/grappler/costs:graph_properties
//tensorflow/core:core_cpu_base
//tensorflow/core:functional_ops_op_lib

Would you help to fix that?

aaroey committed:
Remove unused targets

aaroey committed:
Run clang-format on c++ files

jjsjann123 commented:
Verified that the patch from Benoitsteiner is indeed in the code (making op_performance_data header only). 

Do remember hearing from him about the fix and another patch that is supposed to fix the dependency on grappler/clusters:single_machine.
Maybe the second patch broke the first one ? Should we ping Benoitsteiner, unless @samikama had a fix for this.

jjsjann123 commented:
If the fix is going to take some time, I could revert the code back and use grappler from python side. That would get rid of the dependency. @aaroey

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -0,0 +1,1612 @@
+  // Build the engine
+  trt_builder->setMaxBatchSize(max_batch_size);
+  trt_builder->setMaxWorkspaceSize(max_workspace_size);
+  LOG(INFO) << "starting build engine";
```
Please use VLOG to replace all LOG(INFO) occur hereafter in this file, as I think they're for debugging purpose only.

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -0,0 +1,1612 @@
+    }
+  }
+  // topological order is needed to build TRT network
+  VLOG(-1) << "BUILDING 1";
```
Please remove all logs like this in the file.

samikama committed:
Fix dependency problem

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.h:
```diff
@@ -0,0 +1,52 @@
+        input_inds,  // {node_id, output_idx}
+    const std::vector<std::pair<int, int>>&
+        output_inds,  // {node_id, output_idx}
+    size_t max_batch_size, size_t max_workspace_size,
```
Please follow @wujingyue's suggestion: if it represents the number of bytes, add a '_bytes' suffix to max_workspace_size so user know this is for bytes (instead of KB/MB/etc).

samikama committed:
Merge branch 'tensorrt' of github.com:samikama/tensorflow into tensorrt

samikama commented:
@jjsjann123 I fixed it and testing it locally after merging with changes from @aaroey. Will push it shortly

samikama committed:
Changes for more PR comments

jjsjann123 committed:
[COMMENT]
  trt_engine_op compute
    comment on the scope of pointer arrays for i/o binding
    tensorflow/contrib/tensorrt/kernels/trt_engine_op.cc

samikama committed:
Add missing tf_copts() to Build file so #if GOOGLE_CUDA directives work correctly

samikama commented:
@zheng-xq @aaroey ,
I changed logging statements to VLOG but it seems VLOG logic is inverted. In order to see WARNING messages, INFO messages are also printed, to see FATALs all levels are displayed. 
`
#define VLOG_IS_ON(lvl) \
  ((lvl) <= ::tensorflow::internal::LogMessage::MinVLogLevel())
#endif
`
This is counter-intuitive for me, also inverted logic wrt LOG. See below,
`
LogMessage::~LogMessage() {
  // Read the min log level once during the first call to logging.
  static int64 min_log_level = MinLogLevelFromEnv();
  if (TF_PREDICT_TRUE(severity_ >= min_log_level)) GenerateLogMessage();
}
`
Is this on purpose or is this a bug?

Should I change it? If this is on purpose we need to invert VLOG severities in the code.  

Thanks,
Sami

evanluyifan commented:
Hi @cliffwoolley , I have a simple question, does this change including the tensorRT optimizer? normally tensorRT save the optimized graph as a plan, while in this PR, I think the engine plan is stored in the cache, right?
So, could you show me where we do the optimize job and create a plan for trt_engine_op to load?

aaroey commented:
@samikama That is intended. For VLOG(<level>), higher number of <level> indicates a chattier message. We should not treat <level> the same as INFO/WRANING/etc for LOG, they're different things. Similarly, for LOG(<severity>), don't treat <severity> as a number.

aaroey commented on tensorflow/contrib/tensorrt/README.md:
```diff
@@ -0,0 +1,40 @@
+    gdef, #original graph_def
+    ["output"], #name of output node(s)
+    max_batch_size, #maximum batch size to run the inference
+    max_workspace_size) # max memory for TensorRT to use
```
Please change all occurrence of max_workspace_size to max_workspace_size_bytes.

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -0,0 +1,1603 @@
+  // Build the engine
+  trt_builder->setMaxBatchSize(max_batch_size);
+  trt_builder->setMaxWorkspaceSize(max_workspace_size);
+  LOG(INFO) << "starting build engine "<<static_id;
```
Nit: please add space before and after the double angle bracket.
And please rerun clang-format before commit.

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -0,0 +1,1603 @@
+        infer_object(trt_builder->buildCudaEngine(*converter.network()));
+    LOG(INFO) << "built network";
+    auto engine_plan = infer_object(trt_engine->serialize());
+    VLOG(INFO) << "serialized engine";
```
Please use a number instead, as VLOG levels are different from LOG severities.

aaroey commented on tensorflow/contrib/tensorrt/log/trt_logger.cc:
```diff
@@ -0,0 +1,57 @@
+  // Suppress info-level messages
+  switch (severity) {
+    case Severity::kINFO: {  // Mark TRT info messages as debug!
+      VLOG(-1) << msg;
```
This is effectively LOG(INFO), so please use LOG(INFO) instead to make it consistent with the following.

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -0,0 +1,1603 @@
+    int left = p / 2;
+    int right = p - left;
+
+    VLOG(-1) << "PADDING_" << i << " pre: " << left << ", post: " << right
```
Please use a number larger than 0, as the default vlog level is 0, meaning that you're actually doing a LOG(INFO) here. If you want to print the log, use environment variable [TF_CPP_MIN_VLOG_LEVEL](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/default/logging.cc#L124) instead.

Also, please fix all occurrence of VLOG(-1) in this PR.

aaroey commented on tensorflow/contrib/tensorrt/BUILD:
```diff
@@ -32,6 +45,192 @@ tf_cuda_cc_test(
+    name = "wrap_conversion",
+    srcs = ["trt_conversion.i"],
+    copts = tf_copts(),
+    deps = [
```
On my machine it runs into error like:
```
>>> import tensorflow.contrib.tensorrt
...
ImportError: dynamic module does not define module export function (PyInit__wrap_conversion)
```
Have you encountered similar problem?

samikama replied:
@aaroey,
do you have the lines https://github.com/samikama/tensorflow/blob/3e0396be541f185acbfeaf170ff46bccda0c21eb/tensorflow/tensorflow.bzl#L1299 
in your tensorflow.bzl 
Original rule added ld script that filtered out all init functions if module name doesn't contain tensorflow. Lines above add init_<module_name> function to that list.
Nevertheless I am checking it now with a clean checkout and build

aaroey replied:
I actually have that but there is still import error.

samikama committed:
Change VLOGs to increasing verbosity mode

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.h:
```diff
@@ -38,7 +38,7 @@ tensorflow::Status ConvertSubGraphToTensorRTNodeDef(
     const std::vector<std::pair<int, int>>&
         output_inds,  // {node_id, output_idx}
-    size_t max_batch_size, size_t max_workspace_size_bytes,
+    size_t max_batch_size, size_t max_workspace_size_bytes_bytes,
```
Please revert this :)

jjsjann123 commented:
Just sharing the information. Build is still broken for me.

OpKernel registration seems to have problem. Inference sess run complains about No OpKernel was registered for 'TRTEngineOp'. Sami is looking at it now.

samikama committed:
Fix missing defines in compilation flags that caused missing ops and add a simple test to validate functionality

samikama commented:
Hi,

I just pushed a new commit that fixes the issue with the missing op kernel. It looks like new defines left out the implementation and registration of the op during the compilation due to missing defines in compiler command line. I also added a small test script to validate the functionality of the workflow.

@aaroey  and @wujingyue, if you are happy with the changes, I think we can continue to the next step.

Thanks,
Sami

aaroey commented:
@samikama Thanks for the fix! And sorry about the missing tf_copts(), I guess I would have found the problem earlier if I built it with python 2.7, but I was building with python 3.5 and met that import error mentioned above.

Now I can build, import and run your testing scripts with python 2.7, but still fail with python 3.5 (have you tried that?). I think we can go ahead and merge with upstream first, and I'll add more comments after that.

aaroey committed:
Add/fix copyright and fix format for the test script and other files.

samikama commented:
@aaroey Oops! I forgot that module init functions changed signature between py2 and py3, I fixed that now and testing. Give me a few minutes to do a clean build and I will push it shortly

samikama committed:
Fix ldscript for exporting PyInit_* functions for python3

samikama commented:
Hi @aaroey,

I pushed the updated script. You should be able to import the module from py3 now. However there is a swig problem that prevents use of the module. TF swig bindings seem somehow inconsistent while passing between py->c++->py layers in py3. It looks like std::strings require byte objects for py3->c++ layer on the other hand strings are converted to unicode string when going back from c++ to py3

I will try to see how it is resolved elsewhere in TF code, but I think as you suggested we can proceed as it is right now.

wujingyue commented on tensorflow/contrib/tensorrt/convert/convert_graph.cc:
```diff
@@ -0,0 +1,284 @@
+  cluster =
+      new tensorflow::grappler::VirtualCluster({{"/GPU:0", device_properties}});
+
+  tensorflow::Status status = optimizer.Optimize(cluster, item, &gdef);
```
TF_RETURN_IF_ERROR

jjsjann123 replied:
done

jjsjann123 committed:
[update]
  addressing comment issues

jjsjann123 commented on tensorflow/contrib/tensorrt/convert/convert_nodes.h:
```diff
@@ -0,0 +1,52 @@
+        input_inds,  // {node_id, output_idx}
+    const std::vector<std::pair<int, int>>&
+        output_inds,  // {node_id, output_idx}
+    size_t max_batch_size, size_t max_workspace_size_bytes_bytes,
```
done

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.h:
```diff
@@ -38,7 +38,7 @@ tensorflow::Status ConvertSubGraphToTensorRTNodeDef(
     const std::vector<std::pair<int, int>>&
         output_inds,  // {node_id, output_idx}
-    size_t max_batch_size, size_t max_workspace_size_bytes_bytes,
+    size_t max_batch_size_bytes, size_t max_workspace_size_bytes,
```
Revert the change to max_batch_size? I think it has no units.

jjsjann123 replied:
Oops :)

jjsjann123 committed:
[update]
  addressing comment on variable names

samikama commented on tensorflow/contrib/tensorrt/test/test_tftrt.py:
```diff
@@ -0,0 +1,70 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Script to test TF-TensorRT integration."""
```
Sure @aaroey,

Would it be better to prepare a separate PR for this? This script is not meant to be used as an automated test but a interactive test to see if everything functioning as expected. Proper testing would be more complicated. I think we can do another PR for test version of this and add subsequent tests in later PRs. Does this sound good?

aaroey replied:
Sure. Let's exclude it in this PR then.

samikama committed:
Fix py3 conversion

samikama commented:
Hi @aaroey,

I just pushed changes to make conversion to work both on py3 and py2 so we are in a good shape. If I look above some tests seem to be waiting. How does it work? Is somebody from your side need to trigger it?

aaroey commented on tensorflow/tensorflow.bzl:
```diff
@@ -1296,6 +1296,45 @@ def tf_extension_linkopts():
+# This function attempts to append init_module_name to list of
+# exported functions in version script
+def _append_init_to_versionscript_impl(ctx):
+    modName=ctx.attr.module_name
```
Please fix the naming and format of the newly added code, according to the [bzl style guide](https://docs.bazel.build/versions/master/skylark/bzl-style.html).

aaroey commented:
@samikama I've triggered some of the internal tests, it'll take some time and there will be more later.

jjsjann123 commented on tensorflow/contrib/tensorrt/segment/segment.cc:
```diff
@@ -0,0 +1,262 @@
+  if (VLOG_IS_ON(2)) {
+    for (tensorflow::Node* node : graph.nodes()) {
+      if ((node->in_edges().size() == 0) && (node->out_edges().size() == 0)) {
+        graph.RemoveNode(node);
```
We have since refactored this part.

Currently graph is only used by segment.cc We could remove this function. graph will go out of scope after SegmentGraph.

jjsjann123 committed:
[update]
  removing commented code & unused code

samikama committed:
Fixes for PR comments and build failures

samikama commented:
@aaroey, Thanks triggering tests. We applied the requested changes but since they are outstanding, builds are on hold again. If you are happy with the changes, could you please enable internal tests again so that if there are any problems with the internal checks we can fix them as soon as possible.

samikama committed:
Fix issues with the tests and add mistakenly taken out macro back

samikama commented:
Sorry @aaroey, it looks like kokoro is removing force-run once the run is started. I pushed some fixes for FAILs in sanity tests but they are not run again. Is there any other flag to have them automatically run or do you need to add force-run every time?

aaroey commented:
@samikama Yes I think it requires manual triggering.

samikama commented:
@aaroey it looks like failures are not really related to this or. What do we do in this case

aaroey commented:
@samikama I will try to fix the test today.

aaroey commented on third_party/tensorrt/LICENSE:
```diff
@@ -0,0 +1,203 @@
+Copyright 2018 The TensorFlow Authors.  All rights reserved.
```
I'll do some change internally to add this file since it conflicts with some internal config, and remove it from this PR. Just FYI.

samikama replied:
@aaroey, sure go ahead.

aaroey committed:
Remove LICENSE file. It will be added internally instead.

aaroey committed:
Merge branch 'master' of https://github.com/tensorflow/tensorflow into tensorrt to fix some of the failed tests.

dmikushin commented:
Failed to build for me with the following error:

```
ERROR: /home/dmikushin/.cache/bazel/_bazel_dmikushin/2e6790ed2400dd10dbcc3c8a7729de12/external/nccl_archive/BUILD:33:1: error while parsing .d file: /home/dmikushin/.cache/bazel/_bazel_dmikushin/2e6790ed2400dd10dbcc3c8a7729de12/execroot/org_tensorflow/bazel-out/k8-opt/bin/external/nccl_archive/_objs/nccl/external/nccl_archive/src/core.cu.pic.d (No such file or directory)
nvcc fatal   : Unknown option 'iquote'
nvcc fatal   : Unknown option 'iquote'
```

Configured to compile with nvcc/gcc. I guess -iquote comes from clang

aaroey committed:
Fix some BUILD files

aaroey committed:
Revert "Remove LICENSE file. It will be added internally instead."

This reverts commit d0c2b84c2d6cd3c235671e67f22e3639b4b2b54c.

dmikushin commented:
Still the same problem:

```
ERROR: /home/dmikushin/.cache/bazel/_bazel_dmikushin/2e6790ed2400dd10dbcc3c8a7729de12/external/nccl_archive/BUILD:33:1: error while parsing .d file: /home/dmikushin/.cache/bazel/_bazel_dmikushin/2e6790ed2400dd10dbcc3c8a7729de12/execroot/org_tensorflow/bazel-out/k8-opt/bin/external/nccl_archive/_objs/nccl/external/nccl_archive/src/libwrap.cu.pic.d (No such file or directory)
nvcc fatal   : Unknown option 'iquote'
nvcc fatal   : Unknown option 'iquote'
Target //tensorflow/tools/pip_package:build_pip_package failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 111.376s, Critical Path: 19.96s
FAILED: Build did NOT complete successfully
```

samikama commented:
Hi @dmikushin ,

I am not sure the error you report is related with this PR. You might want to report it upstream. It is an external package not used or modified in this PR.

dmikushin commented:
@samikama This error appears only when I try to compile your PR tensorrt branch.

samikama commented:
@dmikushin, could you please tell me which master commit you tried and what is your platform compiler etc?

samikama commented:
@aaroey, please let us know if we can help with anything in going forward

aaroey commented:
@samikama I'm pulling in this PR internally to make sure it works, there seems to be some swig problem but will let you know if I cannot figure it out.

samikama commented:
@aaroey, please do. Swig interfacing is a nightmare and I wasted hours. Let me know if I can investigate as well, I am happy to look into it.

dmikushin commented:
@samikama This is 37bcd5af97c822e99f617813269351447c79f161 on Ubuntu 17.04, CUDA 9. The build of 0464602ee99ff2a10336f2bea12440167e2d8a70 was previously successful. The '-iquote' problem occurs on random package, actually:

```
ERROR: /home/marcusmae/apc/tensorflow/tensorflow/contrib/rnn/BUILD:218:1: error while parsing .d file: /home/marcusmae/.cache/bazel/_bazel_marcusmae/c5131a6e5cab73016b76ace3066ceaf4/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/contrib/rnn/_objs/python/ops/_lstm_ops_gpu/tensorflow/contrib/rnn/kernels/lstm_ops_gpu.cu.pic.d (No such file or directory)
nvcc fatal   : Unknown option 'iquote'
nvcc fatal   : Unknown option 'iquote'
```

aaroey committed:
1. Use tensorflow::string instead of std::string
2. Fix swig problem in trt_conversion.i
3. Fix c++ styles
4. Fix BUILD dependencies

aaroey commented:
@samikama @jjsjann123  I've made some changes to make sure the code is compatible internally, please help to make sure they work from your side. Thanks.

samikama commented:
@aaroey unfortunately, I am getting a core dump after the changes. I am investigating.

samikama commented:
@aaroey last batch of changes seem to be causing a lot of trouble, would it be possible to revert them? If needed we can replace std::string in interface layers with tensorflow::string and push that change to test you internally.

aaroey committed:
Fix the segfault in convert_nodes.cc

aaroey commented:
@samikama I just pushed a fix, the test script works well on building the trt op for both python2 and 3, but then it detect a mem leak as:
```
2018-02-07 09:22:29.444379: I tensorflow/contrib/tensorrt/convert/convert_nodes.cc:1650] OK finished op building
Traceback (most recent call last):
  File ".../tensorflow/contrib/tensorrt/test/test_tftrt.py", line 67, in <module>
    inpDims[0])  # Get optimized graph
  File ".../tensorflow/contrib/tensorrt/python/trt_convert.py", line 72, in CreateInferenceGraph
    output_graph_def_string = to_bytes(out[1])
  File ".../tensorflow/contrib/tensorrt/python/trt_convert.py", line 49, in py3bytes
    return inp.encode('utf-8', errors='surrogateescape')
AttributeError: 'SwigPyObject' object has no attribute 'encode'
swig/python detected a memory leak of type 'string *', no destructor found.
swig/python detected a memory leak of type 'string *', no destructor found.
```
For python 2 the error is like:
```
2018-02-07 09:44:28.066669: I tensorflow/contrib/tensorrt/convert/convert_nodes.cc:1650] OK finished op building
Traceback (most recent call last):
  File ".../tensorflow/contrib/tensorrt/test/test_tftrt.py", line 67, in <module>
    inpDims[0])  # Get optimized graph
  File ".../tensorflow/contrib/tensorrt/python/trt_convert.py", line 74, in CreateInferenceGraph
    if len(status) < 2:
TypeError: object of type 'SwigPyObject' has no len()
swig/python detected a memory leak of type 'string *', no destructor found.
swig/python detected a memory leak of type 'string *', no destructor found.
```
Any ideas?

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -0,0 +1,1660 @@
+        type_(type),
+        values_(values),
+        owned_values_(owned_values ? *owned_values : std::vector<char>({})),
+        dummy_flag_(false) {
```
@jjsjann123 What does this dummy_flag_ do? Maybe give it a meaningful name?

jjsjann123 replied:
dummy flag means it's an empty weight and the wrapper TRT_ShapedWeights should not construct TRT weights out itself.
Will rename it to empty_weight_flag.

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -0,0 +1,1660 @@
+  }
+
+ private:
+  nvinfer1::ITensor* _tensor_;
```
Please fix all namings in this file, non-static class data member names should start with [a-z] and end with underscore.

jjsjann123 replied:
Working on it.. :)

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -0,0 +1,1660 @@
+  }
+
+  VLOG(2) << "Finished output";
+  static int static_id = 0;
```
This is not thread-safe, non-deterministic, and make test and future debugging difficult. I think it's fine for now, maybe add a TODO to fix it?

jjsjann123 replied:
This is a temporary thing here.
We would refactor to use Tensorflow naming mechanics in the future. Will put TODO here.

samikama replied:
Hi @aaroey,

Conversion code is not meant to be run multi-threaded. Modifying graph multi-threaded would be notoriously hard even if tensorflow::Graph supported it.

aaroey replied:
Hi @samikama,

You're right that in this PR we don't need that; I think we need it once quantization features are implemented after which the conversion may be done during execution instead of graph construction.

samikama commented:
@aaroey unfortunately latest commit is not working for me due to swig bindings. What you see is also indicating that swig bindings are incomplete and possibly has lifetime issues.

aaroey committed:
Fix #includes and namespace scope, and add missing TF_RETURN_IF_ERROR, to make internal build happy.

aaroey commented on tensorflow/contrib/tensorrt/trt_conversion.i:
```diff
@@ -0,0 +1,98 @@
+limitations under the License.
+==============================================================================*/
+
+/* Wrap trt_conversion */
```
TODO: Fix the swig problem. I added this comment to prevent other TF authors from accidentally merging this PR.

samikama replied:
@aaroey, I fixed the swig interface problems but now we see difference between converted and native model. Something else seem to be broken as well. Investigating.

samikama committed:
Fix swig interface problems

jjsjann123 committed:
[update]
  fix naming stuff in tensorflow/contrib/tensorrt/convert/convert_nodes.cc

jjsjann123 committed:
[debug]
  convert const data goes out of scope.

jjsjann123 commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -1128,18 +1128,8 @@ tensorflow::Status ConvertConst(Converter& ctx,
     VLOG(2) << "TENSOR!!!" << node_def.name();
     const auto& content = weights_tensor.tensor_content();
-
-    std::vector<char> values;
```
Doesn't this value go out of scope right after this function?
Later when we get the value into TRT, it's pointing at garbage memory.

Why would we want to copy the memory out? TF owns the weights until the very end of the conversion, while TRT reads from and never writes to it during conversion. It seems safe.

aaroey replied:
You're right that it will go out of scope, but before that in TRT_ShapedWeights we made a copy of the value (line 1142). Actually I'm not sure how memory management works in TRT, so I made a copy every time when a new TRT_ShapedWeights is copied from an old one (see the copy constructor). Is it enough?

And you're right that theoretically we don't need the copy. The main problem is, internally weights_tensor.tensor_content() is not a string, it's a class of complete different type having complete different APIs compared to tensorflow::string, so calling content.data() is invalid and won't compile. The only way to work around this is to call port::CopyToArray(), which internally accepts a different type of input (the internal type of 'content').

It's not ideal but that's the way how TF works. :( So please revert the changes here.

jjsjann123 replied:
Missed the part earlier, thought that owned_value was a pointer. TRT requires only that the data pointer is valid when we construct the network. So keeping a copy should work.
It must be one of the copy we are making that went wrong. The old code is not giving us wrong output from the converted network. Back onto investigation...

jjsjann123 replied:
There's some memory issue with the owned_value_. Data gets corrupted among many copies there. Moreover, we shouldn't need to make more than one copy of each weight anyway.
We allocate temporary memory per weight during the whole scope of the conversion. And pass pointers around.

Removed the std::vector for that part and port::CopyToArray throw data into our temporary memory. Please let me know if it makes internal build happy :)

Next PR will remove owned_value and clean it up.

aaroey replied:
Thanks for the fix! I didn't notice that there was an get_temp_weights() method. Please go ahead to clean it up, as long as we're using CopyToArray it should be fine.

jjsjann123 replied:
Done :)

jjsjann123 committed:
[update]
  Avoid access of data to comply to internal build

jjsjann123 committed:
[clang-format]

samikama commented:
@aaroey,

If you are happy, could we start build tests? Meanwhile you can test for internal compliance.

Thanks,
Sami

jjsjann123 committed:
[cleanup]
  remove TRT_ShapedWeights owned_values_
    fallback to converter temporary memory allocation to avoid redundant memcpy

aaroey commented:
@samikama Sure, I'm building it internally. There is still some internal isssues and I'm trying to fixing them.

samikama commented:
@aaroey is there a way to test us the changes for your internals? Testing and modifying the code for internal and external independently leading to a lot of problems. We have other PRs waiting in the pipeline for this to be merged.

samikama committed:
Merge branch 'master' into tensorrt

aaroey commented:
@samikama Currently it's broken internally but not externally, and unfortunately I think we need to fix the internal issue first. My apologies about the errors from the last few commits, and for some of them like the issues of swig and convert_nodes.cc I was not able to provide correct solutions. Thanks so much for fixing them, and I'll make sure new changes work both internally and externally before I push them.

samikama commented:
@aaroey, errors are not a problem. Such things happen. We are a bit on tight schedule and holding  patches/features for next PRs. This PR has diverged enough that it may take some time to merge and validate further PRs and I am worried that we might be in a tough situation soon. We will try to provide minimalistic (O(100)LOC) PRs next time but we can't proceed without this one. Please let us know if there is anything that we can help. I am open to phone meeting if it helps.

aaroey commented:
@samikama Understood. For python 3 I got the following problem now:
```
2018-02-08 22:35:58.512579: I tensorflow/contrib/tensorrt/convert/convert_nodes.cc:1590] OK finished op building
Traceback (most recent call last):
  File ".../tensorflow/contrib/tensorrt/test/test_tftrt.py", line 67, in <module>
    inpDims[0])  # Get optimized graph
  File ".../tensorflow/contrib/tensorrt/python/trt_convert.py", line 72, in CreateInferenceGraph
    output_graph_def_string = to_bytes(out[1])
  File ".../tensorflow/contrib/tensorrt/python/trt_convert.py", line 49, in py3bytes
    return inp.encode('utf-8', errors='surrogateescape')
AttributeError: 'bytes' object has no attribute 'encode'
```
For python 2 when I tried to `import tensorflow.contrib.tensorrt` it says libnvinfer not found, like:
```
>>> import tensorflow.contrib.tensorrt
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File ".../tensorflow/contrib/tensorrt/__init__.py", line 19, in <module>
    from tensorflow.contrib.tensorrt.python import *
  File ".../tensorflow/contrib/tensorrt/python/__init__.py", line 6, in <module>
    from tensorflow.contrib.tensorrt.python.ops import trt_engine_op
  File ".../tensorflow/contrib/tensorrt/python/ops/trt_engine_op.py", line 31, in <module>
    resource_loader.get_path_to_datafile("_trt_engine_op.so"))
  File ".../tensorflow/contrib/util/loader.py", line 55, in load_op_library
    ret = load_library.load_op_library(path)
  File ".../tensorflow/python/framework/load_library.py", line 58, in load_op_library
    lib_handle = py_tf.TF_LoadLibrary(library_filename, status)
  File ".../tensorflow/python/framework/errors_impl.py", line 516, in __exit__
    c_api.TF_GetCode(self.status.status))
tensorflow.python.framework.errors_impl.NotFoundError: libnvinfer.so.4: cannot open shared object file: No such file or directory
```

samikama commented:
I need to fix #1. For #2 you need tensorrt installation in your LD_LIBRARY_PATH.

aaroey commented:
@samikama Thanks. 2) works now.

I would like to share with you current status of internal build. `tf_gen_op_wrapper_py(name = "trt_engine_op", xxx)` has building errors:
```
.../tensorflow/contrib/tensorrt/gen_trt_engine_op_py_wrappers_cc: error while loading shared libraries: libnvinfer.so.4: cannot open shared object file: No such file or directory
```
It'll take me some time to figure out why this works externally but not internally. This also cause failure of another two `py_library` named `trt_ops_py` and `init_py`. All other targets in contrib/tensorrt work well. So I'm wondering if we can split the failing targets out from this PR and submit the healthy ones first, and send another PR for those targets. In this way you can send future PRs in parallel (assuming you don't depend on the failing targets), and at least we can make some progress.

What do you think?

aaroey committed:
Make build happy by adding missing OP_REQUIRES_OK and if_tensorrt

aaroey committed:
Fix more build dependencies and includes.

samikama commented:
@aaroey, package will not be functional without mentioned targets. The error message you get suggests that the tensorrt is not installed in your build system or glue package in your internal repo is not working as it should. Bazel uses soname to link against concrete library rather than softlink. Could you please verify soname of the library you have is libnvinfer.so.4 and genrule correctly adds this. Alternative could be modifying the genrule of the local_repository generated by tensorrt_configure correctly links them into build area?

aaroey commented:
Thanks @samikama for the suggestions. The package is indeed installed internally and shown in the xx.runfiles directory in bazel-out/, and does have the correct soname. The problem is internally the genrule doesn't provide the same bash environment (more specifically LD_LIBRARY_PATH) as provided by bazel externally, so all cuda shared libraries are not accessible by the gen_trt_engine_op_py_wrappers_cc binary. I think it must be a problem of either the build config or  the internal bazel config, I'll need to figure it out.

aaroey committed:
Fix LINT.IfChange comment

samikama commented:
@aaroey, LD_LIBRARY_PATH is only relevant when running not linking. you can use -s flag to bazel to see exact commandline executed and check whether the paths are constructed correctly or not.

samikama commented:
@aaroey my tests are failing again. Perhaps we should sit together and fix these issues in front of a computer, talking.

aaroey commented:
@samikama sure, set up a meeting, please join at your convenience.

aaroey committed:
Fix python lint errors internally. One important change is to rename CreateInferenceGraph to create_inference_graph.

aaroey committed:
Fix segment_test and run clang-format on trt_conversion.i.

aaroey commented:
@samikama I've figured out the problem of gen_trt_engine_op_py_wrappers_cc. The failure was caused by some internal config, I was able to make it pass under very specific config. So the problem was not related to TF nor this PR, and is not a blocker of the merge.

But we still have swig problems, and we can discuss when we meet.

aaroey committed:
Fix linter errors in test_tftrt.py

samikama committed:
Fix Py3 byte and string issue after swig update. Clarify failure message on finding libnvinfer in configure.py

samikama committed:
Fix merge issues

samikama committed:
Make buildifier happy

tadeegan commented:
Very interesting PR.  I have a few questions..

1) Will this support int8 quantization?
2) What is the difference in strategy between this and the UFF parser / converter Nvidia makes?  This doesn't seem to reuse that project so TensorFlow conversion code will diverge / have differences in behavior between tf.contrib.tensorrt and uff. The ops that this supports is very limited are there more in the works? The UFF parser supports these:
```
    "Add",
    "AvgPool",
    "BiasAdd",
    "ConcatV2",
    "Const",
    "Conv2D",
    "Conv2DBackpropInput",
    "DepthwiseConv2dNative",
    "Div",
    "FusedBatchNorm",
    "MatMul",
    "Maximum",
    "MaxPool",
    "Minimum",
    "Mul",
    "Pack",
    "Placeholder",
    "Relu",
    "Reshape",
    "Shape",
    "Sigmoid",
    "Softmax",
    "StridedSlice",
    "Sub",
    "Tanh",
    "Transpose",
    # https://devtalk.nvidia.com/default/topic/1026657/jetson-tx2/unary-layer-in-tensorrt-3-0-rc/post/5234281/
    # Unary ops don't currently work.
    # "Negative",
    # "Abs",
    # "Sqrt",
    # "Rsqrt",
    # "Pow",
    # "Exp"
    # "Log"
    # Ops that use reduce are currently buggy
    # "Sum",
    # "Prod",
    # "Min",
    # "Max",
    # "Mean",
```

samikama committed:
Don't use tf as directly and import individual modules for internal builds

samikama committed:
fix import

samikama committed:
fix typo

samikama committed:
fix typo

aaroey committed:
Fix test_tftrt.py format

luotao1 commented:
> But there may be some potential performance overhead between the switch of TensorRT of TF(around several hundred us to 1ms observed in our scenarios)

@yangjunpro We are interested in integrating TensorRT with TensorFlow. May I know all the elapsed time of your scenarios, or how much percentage of switch overhead in whole elapsed time? Thanks very much!

aaroey commented:
@luotao1 @yangjunpro Besides the native integration of TensorRT with TF, we're also working on integrating TensorRT to official models and other places, and 
 you can get some numbers [here](https://github.com/tensorflow/models/tree/master/research/tensorrt). Just FYI.

luotao1 commented:
@aaroey Thanks very much!
